### PR TITLE
Make Jenkins builds more robust

### PR DESCRIPTION
**Pull request by dbnicholson**
Try harder to avoid p4a's caching issues and restore normal ccache usage to get back some of the time lost. After this, the cached data is:

* p4a downloaded packages
* ccache objects
* gradle objects

I trust `ccache` and `gradle` to DTRT and I've never seen any issues with p4a's download caching.

https://phabricator.endlessm.com/T33880

### CODE REVIEW

**Comment by dbnicholson** [comment]
The first build with no cache took 36 minutes. Now to try again and see how some caching.

**Comment by dbnicholson** [comment]
retest this please

**Comment by dbnicholson** [comment]
Hmm, the second build took 26 minutes. So, definitely faster but not nearly as much as I was hoping. Still, at this point I'd rather take the reliably correct build.

File: Jenkinsfile [hunk 1]
```diff
@@ -21,6 +13,10 @@ pipeline {
         // FIXME: It would be nice to cache this somehow.
         // FIXME: Go back to use an official release once that happens on GitHub for 0.16.
         KOLIBRI_WHL_URL = 'https://buildkite.com/organizations/learningequality/pipelines/kolibri-python-package/builds/6198/jobs/0182fefa-da30-4683-a144-e7a7c0d3067c/artifacts/0182ff0f-a9e4-4384-bf26-d941243db21c'
+
+        // Both p4a and gradle cache outputs in the home directory.
+        // Point it inside the workspace.
+        HOME = "$WORKSPACE/_cache"
```
Comment (manuq) [inline Jenkinsfile:19]: :+1:

File: p4a-recipes/zstandard/__init__.py [hunk 2]
```diff
@@ -13,6 +15,14 @@ class ZStandardRecipe(CompiledComponentsPythonRecipe):
     )
     depends = ["cffi", "setuptools"]
 
+    # Local patches need to be an absolute paths or they won't be found
+    # by patch.
+    #
+    # https://github.com/kivy/python-for-android/issues/2623
```
Comment (manuq) [inline p4a-recipes/zstandard/__init__.py:21]: OK, according to the p4a PR it's closed and completed by another commit. So we should keep an eye on updating p4a and remove this patch.

File: Makefile [hunk 3]
```diff
@@ -11,13 +11,6 @@ ARCH_OPTIONS := $(foreach arch,$(ARCHES),--arch=$(arch))
 
 OSNAME := $(shell uname -s)
 
-# Usually, P4A prepends "ccache" to CC and CXX. However, this causes certain
-# builds to behave unexpectedly, such as inside the zstandard recipe. Instead,
-# we will add /usr/lib/ccache to $PATH so ccache will be used automatically,
-# and set USE_CCACHE=0 so P4A will use compiler commands as usual.
-export USE_CCACHE := 0
-export PATH := /usr/lib/ccache:$(PATH)
```
Comment (manuq) [inline Makefile:19]: :+1:

**Review by manuq — APPROVED** [review]
Looks great and well explained.

**Comment by manuq** [comment]
> Hmm, the second build took 26 minutes. So, definitely faster but not nearly as much as I was hoping. Still, at this point I'd rather take the reliably correct build.

Yes, I think is better to get the cache properly and wait for half an hour.

File: p4a-recipes/zstandard/__init__.py (hunk 2 above)
Comment (dbnicholson) [reply p4a-recipes/zstandard/__init__.py:21]: Yeah, I'm going to do a p4a rebase next and I'll drop this when I get there.

Comment (dbnicholson) [reply p4a-recipes/zstandard/__init__.py:21]: Oh, I forgot that I already backported this to our p4a branch. Oh well, I'll just clean up post rebase.

**Pull request merged**